### PR TITLE
cloud-controller-manager get /healthz to wait for apiserver to be healthy

### LINKED ISSUE #60288: cloud-controller-manager should get /healthz to wait for apiserver to be healthy
**What happened**:

currently cloud-controller-manager use [`restclient.ServerAPIVersions()` ](https://github.com/kubernetes/kubernetes/blob/master/cmd/cloud-controller-manager/app/controllermanager.go#L263-L269) to wait for apiserver to be healthy.
We should deprecate or remove `ServerAPIVersions` and make use of `/healthz` as all other components do.

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

currently cloud-controller-manager use `restclient.ServerAPIVersions()` to wait for apiserver to be healthy.
Remove ServerAPIVersions and make use of /healthz as all other components do.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #60288

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by jhorwit2** [comment]
@hzxuzhonghu have you tested this on a cluster with RBAC?

**Comment by hzxuzhonghu** [comment]
>@hzxuzhonghu have you tested this on a cluster with RBAC?

@jhorwit2 The /healthz is not limited by authorization.

**Comment by jhorwit2** [comment]
@hzxuzhonghu It is actually. I forgot this was address with the `system:discovery` cluster role/binding a while ago. My bad 👼

**Comment by thockin** [comment]
Assign back to me once it is LGTM'ed and ready for approval.

**Comment by hzxuzhonghu** [comment]
Thanks @thockin

File: staging/src/k8s.io/client-go/rest/versions.go [hunk 1]
```diff
@@ -1,88 +0,0 @@
...
-	defaultAPIPath = "/apis"
-)
-
-// TODO: Is this obsoleted by the discovery client?
-
-// ServerAPIVersions returns the GroupVersions supported by the API server.
-// It creates a RESTClient based on the passed in config, but it doesn't rely
-// on the Version and Codec of the config, because it uses AbsPath and
-// takes the raw response.
-func ServerAPIVersions(c *Config) (groupVersions []string, err error) {
```
Comment (sttts) [inline staging/src/k8s.io/client-go/rest/versions.go]: this being remove here surprises. It needs at least a separate commit.

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 2]
```diff
@@ -261,14 +262,21 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, kubeconfig *rest
...
+		if healthStatus != http.StatusOK {
+			glog.Warningf("apiserver isn't healthy yet. Waiting a little while.")
+			return false, nil
+		}
+
+		return true, nil
 	})
 	if err != nil {
-		glog.Fatalf("Failed to get api versions from server: %v", err)
+		glog.Fatalf("Failed to wait apiserver healthy: %v", err)
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: to wait for apiserver being healthy

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 3]
```diff
@@ -261,14 +262,21 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, kubeconfig *rest
...
+		healthStatus := 0
+		result := versionedClient.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
+		if result.Error() != nil {
+			glog.Errorf("Failed to get apiserver /healthz status: %v", result.Error())
+			return false, nil
 		}
-		glog.Errorf("Failed to get api versions from server: %v", err)
-		return false, nil
+		if healthStatus != http.StatusOK {
+			glog.Warningf("apiserver isn't healthy yet. Waiting a little while.")
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: nit: "Apiserver"

File: staging/src/k8s.io/client-go/rest/versions.go (hunk 1 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/client-go/rest/versions.go]: ok

File: cmd/cloud-controller-manager/app/controllermanager.go [hunk 4]
```diff
@@ -261,14 +262,21 @@ func startControllers(c *cloudcontrollerconfig.CompletedConfig, kubeconfig *rest
 	// If apiserver is not running we should wait for some time and fail only then. This is particularly
 	// important when we start apiserver and controller manager at the same time.
 	err = wait.PollImmediate(time.Second, 10*time.Second, func() (bool, error) {
-		if _, err = restclient.ServerAPIVersions(kubeconfig); err == nil {
-			return true, nil
+		healthStatus := 0
```
Comment (sttts) [inline cmd/cloud-controller-manager/app/controllermanager.go]: Do we have the same code in kube-controller-manager? I would like to have this function in cmd/controller-manager as a generic helper.

Comment (hzxuzhonghu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: I will check.

Comment (hzxuzhonghu) [reply cmd/cloud-controller-manager/app/controllermanager.go]: Yes, also in kube-controller-manager.

**Comment by hzxuzhonghu** [comment]
@sttts  all comments addressed.

**Comment by sttts** [comment]
Can you squash the first and 3rd commit?

**Comment by hzxuzhonghu** [comment]
> Can you squash the first and 3rd commit?

Done

**Comment by hzxuzhonghu** [comment]
@thockin for approval

**Comment by sttts** [comment]
/assign @luxas 
for cloud-ctrl-manager.

/assign @deads2k 
for kube-ctrl-manager.

File: cmd/controller-manager/app/helper.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+func WaitForAPIServer(client clientset.Interface, timeout time.Duration) error {
+	return wait.PollImmediate(time.Second, timeout, func() (bool, error) {
+		healthStatus := 0
+		result := client.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
+		if result.Error() != nil {
+			glog.Errorf("Failed to get apiserver /healthz status: %v", result.Error())
+			return false, nil
+		}
+		if healthStatus != http.StatusOK {
+			glog.Warningf("APIServer isn't healthy yet. Waiting a little while.")
```
Comment (deads2k) [inline cmd/controller-manager/app/helper.go]: keep track of the body.  If the entire check fails, return an error that contains the current text of /healthz so the callers will print it when they print the error.

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 6]
```diff
@@ -350,34 +348,15 @@ func NewControllerInitializers(loopMode ControllerLoopMode) map[string]InitFunc
...
-			return false, nil
-		}
-		content, _ := resp.Raw()
-		healthzContent = string(content)
-
-		discoveryClient = client.Discovery()
-		return true, nil
-	})
+	client := clientBuilder.ClientOrDie("controller-discovery")
+	err := genericcontrollerconfig.WaitForAPIServer(client, 10*time.Second)
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/controllermanager.go]: Promote this out into the `CreateContext` method where it is called.  Make it more obvious that we wont'/can't produce a context on a bad API server.

This bit of code about available resources is likely to be a target for refactoring.

**Comment by deads2k** [comment]
I've got a couple comments I'd like cleaned up, but I like the change.

**Comment by thockin** [comment]
@deads2k I'm going to approve and remove @sttts LGTM, so you can LGTM when satisfied.

File: cmd/controller-manager/app/helper.go (hunk 5 above)
Comment (hzxuzhonghu) [reply cmd/controller-manager/app/helper.go]: make sense

File: cmd/kube-controller-manager/app/controllermanager.go (hunk 6 above)
Comment (hzxuzhonghu) [reply cmd/kube-controller-manager/app/controllermanager.go]: good idea

**Comment by hzxuzhonghu** [comment]
@sttts @deads2k comments addressed, PTAL

File: cmd/controller-manager/app/helper.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+// WaitForAPIServer waits for the API Server's /healthz endpoint to report "ok" with timeout.
+func WaitForAPIServer(client clientset.Interface, timeout time.Duration) error {
+	var err error
+
+	pollErr := wait.PollImmediate(time.Second, timeout, func() (bool, error) {
```
Comment (sttts) [inline cmd/controller-manager/app/helper.go]: I would have renamed the err above into lastErr and kept this as err. An `err` variable is by convention a local thing in Golang.

Comment (sttts) [reply cmd/controller-manager/app/helper.go]: it's too easy to accidently shadow `err`.

Comment (hzxuzhonghu) [reply cmd/controller-manager/app/helper.go]: make sense

File: cmd/controller-manager/app/helper.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+			return false, nil
+		}
+		// set err = nil when condition satisfied
+		err = nil
+
+		return true, nil
+	})
+
+	if pollErr != nil {
+		err = fmt.Errorf("%v: %v", pollErr, err)
```
Comment (sttts) [inline cmd/controller-manager/app/helper.go]: return fmt.Errorf(...)

File: cmd/kube-controller-manager/app/controllermanager.go [hunk 9]
```diff
@@ -407,6 +379,13 @@ func CreateControllerContext(s *config.CompletedConfig, rootClientBuilder, clien
 	versionedClient := rootClientBuilder.ClientOrDie("shared-informers")
 	sharedInformers := informers.NewSharedInformerFactory(versionedClient, ResyncPeriod(s)())
 
+	// If apiserver is not running we should wait for some time and fail only then. This is particularly
+	// important when we start apiserver and controller manager at the same time.
+	err := genericcontrollerconfig.WaitForAPIServer(versionedClient, 10*time.Second)
```
Comment (sttts) [inline cmd/kube-controller-manager/app/controllermanager.go]: if err := ...; err != nil {

Comment (hzxuzhonghu) [reply cmd/kube-controller-manager/app/controllermanager.go]: ok

File: cmd/controller-manager/app/helper.go (hunk 8 above)
Comment (hzxuzhonghu) [reply cmd/controller-manager/app/helper.go]: ok

**Comment by sttts** [comment]
@deads2k's comments are addressed.

/lgtm

**Pull request merged**